test/CheckedC/dynamic-checks/array-subscript-code-gen.c

```diff
@@ -0,0 +1,1021 @@
+  // CHECK-AST-NEXT: ImplicitCastExpr {{.*}} <LValueToRValue>
+  // CHECK-AST-NEXT: Inferred Bounds
+  // CHECK-AST-NEXT: RangeBoundsExpr
+  // CHECK-AST-NEXT: ImplicitCastExpr {{.*}} <ArrayToPointerDecay>
+  // CHECK-AST-NEXT: DeclRefExpr {{.*}} 'gma' 'int checked[3][3]'
+  // CHECK-AST-NEXT: BinaryOperator {{.*}} '_Array_ptr<int checked[3]>':'_Array_ptr<int checked[3]>' '+'
+  // CHECK-AST-NEXT: ImplicitCastExpr {{.*}} <ArrayToPointerDecay>
+  // CHECK-AST-NEXT: DeclRefExpr {{.*}} 'gma' 'int checked[3][3]'
+  // CHECK-AST-NEXT: IntegerLiteral {{.*}} 'unsigned long long' 3
+  // CHECK-AST-NEXT: ArraySubscriptExpr
+  // CHECK-AST-NEXT: ImplicitCastExpr {{.*}} <ArrayToPointerDecay>
+  // CHECK-AST-NEXT: ArraySubscriptExpr
+
+  // The following line of LLVM IR is offensively long
+  // CHECK-IR: br i1 and
+  // CHECK-IR-SAME: (i1 icmp sle (i32 ptrtoint ([3 x [3 x i32]]* @gma to i32),
+  // CHECK-IR-SAME: i32 ptrtoint (i32* getelementptr inbounds ([3 x [3 x i32]], [3 x [3 x i32]]* @gma, i32 0, i64 1, i32 2) to i32)),
+  // CHECK-IR-SAME: i1 icmp slt (i32 ptrtoint (i32* getelementptr inbounds ([3 x [3 x i32]], [3 x [3 x i32]]* @gma, i32 0, i64 1, i32 2) to i32),
+  // CHECK-IR-SAME: i32 ptrtoint ([3 x i32]* getelementptr inbounds
+  // CHECK-IR-SAME: ([3 x i32], [3 x i32]* getelementptr inbounds ([3 x [3 x i32]], [3 x [3 x i32]]* @gma, i32 0, i32 0), i32 3) to i32))),
+  // CHECK-IR-SAME: label %[[LAB_DYSUC:_Dynamic_check.succeeded[a-zA-Z0-9.]*]],
+  // CHECK-IR-SAME: label %{{_Dynamic_check.failed[a-zA-Z0-9.]*}}
+  // CHECK-IR: [[LAB_DYSUC]]:
+  // CHECK-IR-NEXT: [[REG1:%[a-zA-Z0-9.]*]] = load i32, i32* getelementptr inbounds ([3 x [3 x i32]], [3 x [3 x i32]]* @gma, i32 0, i64 1, i32 2), align 4
+  // CHECK-IR-NEXT: store i32 [[REG1]], i32* %z3, align 4
+
+  // Also Valid
+  int z4 = gma[2][-2]; // expected-warning {{index -2 is before the beginning of the array}}
+  // CHECK-AST: VarDecl {{.*}} z4 'int' cinit
+  // CHECK-AST-NEXT: ImplicitCastExpr {{.*}} <LValueToRValue>
+  // CHECK-AST-NEXT: Inferred Bounds
+  // CHECK-AST-NEXT: RangeBoundsExpr
+  // CHECK-AST-NEXT: ImplicitCastExpr {{.*}} <ArrayToPointerDecay>
+  // CHECK-AST-NEXT: DeclRefExpr {{.*}} 'gma' 'int checked[3][3]'
+  // CHECK-AST-NEXT: BinaryOperator {{.*}} '_Array_ptr<int checked[3]>':'_Array_ptr<int checked[3]>' '+'
+  // CHECK-AST-NEXT: ImplicitCastExpr {{.*}} <ArrayToPointerDecay>
+  // CHECK-AST-NEXT: DeclRefExpr {{.*}} 'gma' 'int checked[3][3]'
+  // CHECK-AST-NEXT: IntegerLiteral {{.*}} 'unsigned long long' 3
+  // CHECK-AST-NEXT: ArraySubscriptExpr
+  // CHECK-AST-NEXT: ImplicitCastExpr {{.*}} <ArrayToPointerDecay>
+  // CHECK-AST-NEXT: ArraySubscriptExpr
+
+  // The following line of LLVM IR is offensively long
+  // CHECK-IR: br i1 and
+  // CHECK-IR-SAME: (i1 icmp sle (i32 ptrtoint ([3 x [3 x i32]]* @gma to i32),
+  // CHECK-IR-SAME: i32 ptrtoint (i32* getelementptr inbounds ([3 x [3 x i32]], [3 x [3 x i32]]* @gma, i32 0, i32 2, i32 -2) to i32)),
```

Comment:
This indexing is more clearly "correct" as a translation. That said, I think a test that actually runs this code dynamically would be more useful.